library/ctr_drbg.c

```diff
@@ -94,11 +94,15 @@ int mbedtls_ctr_drbg_seed_entropy_len(
     /*
      * Initialize with an empty key
      */
-    mbedtls_aes_setkey_enc( &ctx->aes_ctx, key, MBEDTLS_CTR_DRBG_KEYBITS );
+    if( ( ret = mbedtls_aes_setkey_enc( &ctx->aes_ctx, key, MBEDTLS_CTR_DRBG_KEYBITS ) ) != 0 )
+    {
+        return( ret );
+    }
 
     if( ( ret = mbedtls_ctr_drbg_reseed( ctx, custom, len ) ) != 0 )
+    {
         return( ret );
-
+    }
     return( 0 );
```

Comment:
I think we should add a `cleanup` section here which for `ret != 0` calls `mbedtls_aes_free` on the AES context. Note, though, that this relies on the convention that it should always be safe to call `xxx_free` even if previous calls to `xxx_some_operation` have failed, which I think should be true, but I'm not 100% sure. @gilles-peskine-arm@sbutcher-arm @Patater

Reply:
On the other hand, the AES context in question is not local but part of the CTR DRBG structure, so a subsequent call to `ctr_drbg_free` would also take care of the call to `mbedtls_aes_free`. So, while it's unusual to see a structure that's `xxx_init`ialized in a function but not `xxx_free`d on failure, it should be OK here.

Reply:
@hanno-arm  Our convention is that it's not only always correct, but in fact mandatory to call `xxx_free` on any object on which `xxx_init` has been called and `xxx_free` has not subsequently been called. The fact that `xxx_something` has failed does not necessarily mean that any embedded data structure has been freed and all sensitive data in the context has been cleared. As you note, this function should arguably call `aes_free` to free resources containing sensitive data as soon as possible; however, the caller of this function should react to the failure by calling `ctr_drbg_free` anyway, and that will take care of calling `aes_free`. So that's ok.

Reply:
@gilles-peskine-arm Agreed, thanks for checking!